types/node/stream.d.ts

```diff
@@ -1333,46 +1367,57 @@ declare module "node:stream" {
+            : never;
+        type PipelineTransformGenerator<S extends PipelineTransformSource<any>, O> = (
+            source: PipelineSourceArgument<S>,
+            options?: Abortable,
+        ) => AsyncIterable<O>;
+        type PipelineTransformStreams<I, O> =
             | NodeJS.ReadWriteStream
-            | ((
-                source: S extends (...args: any[]) => Iterable<infer ST> | AsyncIterable<infer ST> ? AsyncIterable<ST>
-                    : S,
-            ) => AsyncIterable<U>);
-        type PipelineTransformSource<T> = PipelineSource<T> | PipelineTransform<any, T>;
-        type PipelineDestinationIterableFunction<T> = (source: AsyncIterable<T>) => AsyncIterable<any>;
-        type PipelineDestinationPromiseFunction<T, P> = (source: AsyncIterable<T>) => Promise<P>;
-        type PipelineDestination<S extends PipelineTransformSource<any>, P> = S extends
-            PipelineTransformSource<infer ST> ?
+            | web.TransformStream<I, O>;
+        type PipelineTransform<S extends PipelineTransformSource<any>, O> = S extends
+            PipelineSource<infer I> | PipelineTransformStreams<any, infer I> | ((...args: any[]) => infer I)
+            ? PipelineTransformStreams<I, O> | PipelineTransformGenerator<S, O>
+            : never;
+        type PipelineTransformSource<O> = PipelineSource<O> | PipelineTransform<any, O>;
+        type PipelineDestinationFunction<S extends PipelineTransformSource<any>, R> = (
+            source: PipelineSourceArgument<S>,
+            options?: Abortable,
+        ) => R;
+        type PipelineDestination<S extends PipelineTransformSource<any>, R> = S extends
+            PipelineSource<infer I> | PipelineTransform<any, infer I> ?
                 | NodeJS.WritableStream
-                | PipelineDestinationIterableFunction<ST>
-                | PipelineDestinationPromiseFunction<ST, P>
+                | web.WritableStream<I>
+                | web.TransformStream<I, any>
+                | PipelineDestinationFunction<S, R>
             : never;
-        type PipelineCallback<S extends PipelineDestination<any, any>> = S extends
-            PipelineDestinationPromiseFunction<any, infer P> ? (err: NodeJS.ErrnoException | null, value: P) => void
-            : (err: NodeJS.ErrnoException | null) => void;
-        type PipelinePromise<S extends PipelineDestination<any, any>> = S extends
-            PipelineDestinationPromiseFunction<any, infer P> ? Promise<P> : Promise<void>;
-        interface PipelineOptions {
-            signal?: AbortSignal | undefined;
-            end?: boolean | undefined;
-        }
+        type PipelineCallback<S extends PipelineDestination<any, any>> = (
+            err: NodeJS.ErrnoException | null,
+            value: S extends (...args: any[]) => PromiseLike<infer R> ? R : undefined,
+        ) => void;
+        type PipelineResult<S extends PipelineDestination<any, any>> = S extends NodeJS.WritableStream ? S : Duplex;
```

Comment:
These changes lend some advantages:
- the return values of generator-style transformers are passed through the pipeline as the original type, without narrowing to `AsyncIterator`
- web streams are permitted, and chunk types are inferred from web streams
- non-stream destinations result correctly in a Duplex

All of the existing tests pass, modulo some intentional ExpectTypes.